crates/core/tedge_agent/src/file_transfer_server/entity_store.rs

```diff
@@ -0,0 +1,409 @@
+    fn into_response(self) -> Response {
+        let status_code = match &self {
+            Error::InvalidEntityTopicId(_) => StatusCode::BAD_REQUEST,
+            Error::EntityStoreError(_) => StatusCode::BAD_REQUEST,
+            Error::EntityNotFound(_) => StatusCode::NOT_FOUND,
+            Error::ChannelError(_) => StatusCode::INTERNAL_SERVER_ERROR,
+            Error::InvalidEntityStoreResponse => StatusCode::INTERNAL_SERVER_ERROR,
+        };
+        let error_message = self.to_string();
+
+        (status_code, error_message).into_response()
+    }
+}
+
+pub(crate) fn entity_store_router(state: AgentState) -> Router {
+    Router::new()
+        .route(
+            "/v1/entities/*path",
+            post(register_entity)
+                .get(get_entity)
+                .delete(deregister_entity),
+        )
+        .with_state(state)
+}
+
+async fn register_entity(
+    State(state): State<AgentState>,
+    Json(entity): Json<EntityRegistrationMessage>,
+) -> Result<StatusCode, Error> {
+    let response = state
+        .entity_store_handle
+        .clone()
+        .await_response(EntityStoreRequest::Create(entity.clone()))
+        .await?;
+    let EntityStoreResponse::Create(result) = response else {
+        return Err(Error::InvalidEntityStoreResponse);
+    };
+
+    if !result?.0.is_empty() {
+        let message = entity.to_mqtt_message(&state.mqtt_schema);
+        state.mqtt_publisher.clone().send(message).await?;
+    }
+    Ok(StatusCode::OK)
+}
+
+async fn get_entity(
+    State(state): State<AgentState>,
+    Path(path): Path<String>,
+) -> Result<Json<EntityMetadata>, Error> {
+    let topic_id = EntityTopicId::from_str(&path)?;
+
+    let response = state
+        .entity_store_handle
+        .clone()
+        .await_response(EntityStoreRequest::Get(topic_id.clone()))
+        .await?;
+
+    let EntityStoreResponse::Get(entity_metadata) = response else {
+        return Err(Error::InvalidEntityStoreResponse);
+    };
+
+    if let Some(entity) = entity_metadata {
+        Ok(Json(entity.clone()))
+    } else {
+        Err(Error::EntityNotFound(topic_id))
+    }
+}
+
+async fn deregister_entity(
+    State(state): State<AgentState>,
+    Path(path): Path<String>,
+) -> Result<StatusCode, Error> {
+    let topic_id = EntityTopicId::from_str(&path)?;
+
+    let response = state
+        .entity_store_handle
+        .clone()
+        .await_response(EntityStoreRequest::Delete(topic_id.clone()))
+        .await?;
+
+    let EntityStoreResponse::Delete(deleted) = response else {
+        return Err(Error::InvalidEntityStoreResponse);
+    };
+
+    for topic_id in deleted {
+        let topic = state
+            .mqtt_schema
+            .topic_for(&topic_id, &Channel::EntityMetadata);
+        let clear_entity_msg = MqttMessage::new(&topic, "")
+            .with_retain()
+            .with_qos(QoS::AtLeastOnce);
+
+        state.mqtt_publisher.clone().send(clear_entity_msg).await?;
+    }
```

Comment:
I would delegate to `EntityStoreServer` the responsibility to publish over MQTT the updated entity metadata.

-  `EntityStoreServer`  is already MQTT dependent and publishing entity metadata.
- This HTTP handler would be free of any MQTT dependencies.

Reply:
Resolved by 1a85933b